### ISSUE TITLE
PDF customizations in 3.0 broken

### ISSUE THREAD

**Issue by shaneataylor** [dita-ot/dita-ot#2835]
## Expected Behavior
Migrated PDF plugin to 3.0 from 2.5.4 should include all customizations.

## Actual Behavior
Most of the customizations are not implemented. 

## Possible Solution
Still investigating. This _could_ be a docs issue if the plugin needs to be changed in a particular way, but I think it might be an OT issue as the release notes don't suggest major changes in the PDF implementation that would cause users to need to migrate.

## Steps to Reproduce
[com.cengage.userguide.zip](https://github.com/dita-ot/dita-ot/files/1444268/com.cengage.userguide.zip)

1. Install attached plugin with DITA-OT 2.5.4 and generate output with the userguide format.
2. Install attached plugin with DITA-OT 3.0 and generate output with the userguide format.

## Environment
* DITA-OT version: 3.0
* Operating system and version: MacOS 10.12.6
* How did you run DITA-OT? dita command
* Transformation type: PDF-based customization

**Comment by ru-fu** [comment]
I'm seeing the same problem. When I install my custom PDF plugin in 2.5.4, it works without problems. When I install it in 3.0.1, it doesn't seem to be used at all (output is standard PDF, no custom messages in the log).
I compared the logs (using preprocess instead of preprocess2) and couldn't find anything that looked suspicious ...

**Comment by raducoravu** [comment]
@shaneataylor your "com.cengage.userguide" plugin depends on "com.cengage.strings" which is not inside the zip so I am unable to test with it.
@ru-fu could you also link to a zip containing a test plugin?

**Comment by T-Spark** [comment]
I experienced the same problem. The plugin works under dita-ot 2.5.1, but is completely ignored under 3.0.1. Find the plugin attached as zip.
[com.novexx.pdf-ETK.zip](https://github.com/dita-ot/dita-ot/files/1673528/com.novexx.pdf-ETK.zip)

**Comment by ru-fu** [comment]
Here's a very basic plugin built according to the instructions in the documentation:
[com.example.print-pdf.zip](https://github.com/dita-ot/dita-ot/files/1673614/com.example.print-pdf.zip)

**Comment by raducoravu** [comment]
Thanks for the samples.
@jelovirt, at a first glance, if in the ANT target "transform.topic2fo.main" from (dita-ot-3.0.1\plugins\org.dita.pdf2\build.xml) we echo the "${xml.catalog.files}" parameter, it contains the proper XML catalogs list (the first being the XML catalog from the PDF customization folder followed by the others).
But still the XSLT processor when solving this xsl:import reference:

           <xsl:import href="cfg:fo/xsl/custom.xsl"/>

will prefer with high priority the mapping from the main "catalog-dita.xml" (which refers to "plugins/org.dita.pdf2/cfg/catalog.xml" which contains a mapping for "cfg:fo/xsl/custom.xsl").

The culprit for this seems to be this file "dita-ot-3.0.1\config\CatalogManager.properties" which is probably picked up and used by the XML Catalog resolver with high priority instead of (or before) the XML catalogs list specified in the <xslt> task.
Renaming the "CatalogManager.properties" to something else seems to again make the PDF customization folder (specified in the PDF customization plugin) take priority.
So we should at least remove that "catalogs=../catalog-dita.xml" mapping from the "CatalogManager.properties". Or remove the "CatalogManager.properties" completely. Why was it added anyway?

**Comment by larrykollar** [comment]
I'm having the same problem. Unfortunately, renaming config/CatalogManager.properties didn't solve the problem for me. Interestingly enough, the log file (and the output) indicates DITA-OT is picking up the values in cfg/build.properties (called out in integrator.xml in my plugin). It's just not grabbing the custom.xsl files called out in the catalog. Another interesting data point: DITA-OT was picking up variable strings defined in a different plugin (a different PDF customization) although there's no reference to that plugin in this one.

Using `blame` on the CatalogManager.properties file suggests that catalogs entry has been there since the beginning.

**Comment by raducoravu** [comment]
👍 We had another end user report this problem:
https://www.oxygenxml.com/forum/post46116.html#p46116

**Comment by merritcat** [comment]
I have the same problem. A plugin that works with dita-ot 2.3.3 is completely ignored in 3.0.3.

``

**Comment by raducoravu** [comment]
How about trying this workaround?
Open the main "catalog-dita.xml" and comment out the reference to:

        <nextCatalog catalog="plugins/org.dita.pdf2/cfg/catalog.xml"/>

The first time the ""org.apache.xml.resolver.CatalogManager" class is instantiated (and it's instantiated only once being a single Java process) it finds the "dita-ot-3.0.3\config\CatalogManager.properties" in the classpath and on the "org.apache.xml.resolver.CatalogManager.queryCatalogFiles()" method it finds a reference to "catalog-dita.xml". This will be the first parsed and used XML catalog. 
All other XML catalogs are added after the "catalog-dita.xml" in the Apache catalog resolver.

And the "catalog-dita.xml" has a reference to:

              <nextCatalog catalog="plugins/org.dita.pdf2/cfg/catalog.xml"/>

so URI references like "cfg:fo/xsl/custom.xsl" will always get resolved to the mappings in the "plugins/org.dita.pdf2/cfg/catalog.xml".

So what changed between DITA OT 2.x and 3.x?
I suspect that now the "dita.bat" adds in the classpath the folder "%DITA_HOME%\config" allowing the Xerces CatalogManager to find the "CatalogManager.properties" and load it.

@robander  One question, why does the ""catalog-dita.xml" " need to have a reference to "plugins/org.dita.pdf2/cfg/catalog.xml"?

**Comment by robander** [comment]
> @robander One question, why does the ""catalog-dita.xml" " need to have a reference to "plugins/org.dita.pdf2/cfg/catalog.xml"?

I'm not sure - I have never touched that :smile: I would have guessed that we're using the same catalog to resolve everything, including doctypes and the `plugin:` or `cfg:` syntax in XSL imports but I have done very little with the `cfg:` imports in the past.

**Comment by larrykollar** [comment]
@raducoravu, commenting out the nextCatalog element didn't make a difference. I did a `dita --install` in case that was needed, but it didn't help.

**Comment by easirois** [comment]
I have a similar issue.  To resolve it I had to rename the files in my plugin so they were not the same as the ones in the base pdf2 plugin. Adjusted my catalog file in the plugin accordingly and everything worked fine. In my case there was an error about duplicate files having already been loaded.  I thought it my have been a Saxon 9.8 issue, but I had the same behavior when using Saxon 9.1 in 3.0.3 as well

**Comment by jelovirt** [comment]
A quick test and I wasn't able to reproduce this. Could someone add a simple plugin that doesn't work with 3.0.

**Comment by ru-fu** [comment]
I tested the basic plugin that I posted again, and it works when I follow @raducoravu's instructions and comment out the pdf2 catalog. Otherwise it doesn't.
I could post another plugin that doesn't work for me, but that probably won't help - so what is different in your setup than in others'? ;)

**Comment by larrykollar** [comment]
After commenting out that line (line 60 in org.dita.pdf2/plugin.xml), then running `dita --install`, my custom plugin worked!

The question is, will there be unwanted side effects from this? If not, I'd suggest yanking the line and releasing a hot fix. :)

**Comment by jelovirt** [comment]
@raducoravu are you able to reproduce this on both OS X and Windows?

**Comment by raducoravu** [comment]
@jelovirt Yes, I can also reproduce it on Mac OSX testing from the command line using something like:

            ./dita -i /Users/raducoravu/Documents/samples/dita/flowers/flowers.ditamap -f print-pdf

Commenting out that reference to the customization XML catalog in '"org.dita.pdf2\plugin.xml"" and running the integrator afterwards again works and leads to the customization being used.
Renaming the "CatalogManager.properties" to some other file name also works to make the PDF customization XSLT start being used.

**Comment by raducoravu** [comment]
@T-Spark Actually this warning:

         [xslt] C:\DITA-OT\DITA-OT-3.0.3\plugins\com.novexx.pdf-Doc_oL\cfg\fo\xsl\front-matter.xsl:82:40: Warning! A variable with no following sibling instructions has no effect
     [xslt] Failed to process C:\Users\LUTZLA~1\AppData\Local\Temp\temp20180416090055420\stage1.xml

means that your XSLT customization is taken into account.
About the failure:

     [xslt] C:\DITA-OT\DITA-OT-3.0.3\plugins\org.dita.pdf2\xsl\fo\topic.xsl:562:69: Fatal Error! In {fn:string-length(...) lt $maxCharsInShortDesc}: cannot compare xs:integer to xs:untypedAtomic

The variable "maxCharsInShortDesc" is defined in the "dita-ot-3.0.3\plugins\org.dita.pdf2\xsl\fo\commons.xsl" as having xs:integer type:

    <xsl:variable name="maxCharsInShortDesc" as="xs:integer">
        <xsl:call-template name="getMaxCharsForShortdescKeep"/>
    </xsl:variable>

so maybe this fails because you have a custom XSLT template which re-defines it without assigning a type to it. 
Basically maybe your PDF customization folder needs to be migrated and modified in order to work correctly with DITA OT 3.x.

**Comment by T-Spark** [comment]
@raducoravu: thanks for your notes. I just added

> as="xs:integer"
to the variable declaration of  "maxCharsInShortDesc" with the result that  my custom plugin works now! (except of some minor problems like not found images and not found hyphenation file)

**Comment by jelovirt** [comment]
@raducoravu could you attach the plugin that has the `print-pdf` you're using.
